Add pdal feature Attempt 1

RealEdwardS opened the pull request:
### Describe your changes

### Issue ticket number and link if any

### Checklist for finalizing the PR

- [✔] I have performed a [self-review](https://f3d.app/doc/dev/CODING_STYLE.html) of my code
- [] I have added [tests](https://f3d.app/doc/dev/TESTING.html) for new features and bugfixes
- [ ] I have added [documentation](https://f3d.app/) for new features
- [ ] If it is a modifying the libf3d API, I have updated bindings
- [ ] If it is a modifying the `.github/workflows/versions.json`, I have updated `timestamp`

### Continuous integration

Please check the checkbox of the CI you want to run, then push again on your branch.

- [ ] Style checks
- [ ] Fast CI
- [ ] Coverage cached CI
- [ ] Analysis cached CI
- [ ] WASM docker CI
- [ ] Android docker CI
- [ ] macOS Intel cached CI
- [ ] macOS ARM cached CI
- [ ] Windows cached CI
- [ ] Linux cached CI
- [ ] Other cached CI

mwestphal commented:
Hi @RealEdwardS 

Looks like this is ongoing, do not hesitate to ping me here or on discord when you need a review :)

RealEdwardS commented:
Fixed and added some material.

Right now I've skipped some supported PDAL types including faux, i3s, memoryview, pgpointcloud, and tindex, since they're not files, but other buffers/mediums? 

The gdal has a lot of files it supports, I still need to add all of them if necessary. 

Hopefully I'm on the right path, if I am, I'll continue and add some testing.

Thanks :)

RealEdwardS commented:
Can you review my current code? I don't think I've missed anything with the implementation, though I haven't touched anything in the .github directory which I assume is used to test and build the plugin.

Thanks

mwestphal commented on plugins/CMakeLists.txt:
```diff
@@ -40,3 +40,12 @@ if (F3D_PLUGIN_BUILD_VDB)
+
+if (F3D_PLUGIN_BUILD_PDAL)
+  # PDAL plugin needs vtkPDALReader: https://gitlab.kitware.com/vtk/vtk/-/tree/master/IO/PDAL
+  if(VTK_VERSION VERSION_GREATER_EQUAL 9.4.0)
```
why though ? vtkPDALReader is present in older version of VTK afair ?

RealEdwardS replied:
Sorry! I was using the most recent version. Updated.

Meakk commented on plugins/pdal/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+
+f3d_plugin_declare_reader(
+    NAME PDAL
+    EXTENSIONS arrow bpf ilvis2 e57 fbi mbio npy csd pcd ply qfit rdb rxp sbet bin tiledb las drc ept.json json hdf h5 mat ntif obj pts ptx txt splk spz
```
Also: `obj`, `ply` and `spz` are handled natively. `drc` is handled by `draco` plugin. `hdf`, `h5` are handled by `hdf` plugin. We need to sort that out.

mwestphal replied:
I agree with @Meakk here.


We do not want to add all these extensions, lets add:

    - The one we actually want : e57 gdal las
    - The one we can test: I leave it up to you. but each of the individual format will need a test.
    -And remove the one that we dont want to use PDAL for: obj, pts, spz, ply, npy,

mwestphal replied:
lets focus solely on e57 for now.

Meakk commented on plugins/pdal/pdal.inl:
```diff
@@ -0,0 +1,5 @@
+void applyCustomReader(vtkAlgorithm* algo, const std::string&) const override
+{
+  vtkPDALReader* pdalReader = vtkPDALReader::SafeDownCast(algo);
+  pdalReader->UpdateInformation();
```
Is it needed?

mwestphal replied:
should not be needed ?

mwestphal commented on plugins/pdal/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+
+f3d_plugin_declare_reader(
+    NAME PDAL
+    EXTENSIONS arrow bpf ilvis2 e57 fbi mbio npy csd pcd ply qfit rdb rxp sbet bin tiledb las drc ept.json json hdf h5 mat ntif obj pts ptx txt splk spz
```
We do not want to add all these extensions, lets add:

1. The one we actually want : e57 gdal las 
2. The one we can test: I leave it up to you. but each of the individual format will need a test.
3. And remove the one that we dont want to use PDAL for: obj, pts, spz, ply, npy,

mwestphal commented:
will fix https://github.com/f3d-app/f3d/issues/201

mwestphal commented:
Hi @RealEdwardS , how is it going ? Do you need any help ?

mwestphal commented:
@RealEdwardS let me know if you have any struggle reaching out on discord

mwestphal commented on .DS_Store:
Please remove this file from your PR

mwestphal commented on .github/actions/pdal-install-dep/action.yml:
```diff
@@ -0,0 +1,76 @@
+        cmake ../pdal
+        -DBUILD_SHARED_LIBS=ON
+        -DCMAKE_BUILD_TYPE=Release
+        -DCMAKE_INSTALL_LIBDIR:PATH=lib
```
do you need that ?

RealEdwardS replied:
Tested it without the file, it works - Removed

mwestphal commented on .github/actions/pdal-install-dep/action.yml:
```diff
@@ -0,0 +1,76 @@
+        -DCMAKE_BUILD_TYPE=Release
+        -DCMAKE_INSTALL_LIBDIR:PATH=lib
+        -DCMAKE_INSTALL_PREFIX=../pdal_install
+        -DCMAKE_POSITION_INDEPENDENT_CODE=ON
```
do you need this ?

mwestphal commented on application/testing/CMakeLists.txt:
```diff
@@ -955,6 +955,10 @@ if(F3D_PLUGIN_BUILD_VDB)
 endif ()
 
+if(F3D_PLUGIN_BUILD_PDAL)
+  f3d_test(NAME TestE57File DATA bunnyInt19.e57 ARGS --load-plugins=pdal)
```
```suggestion
  f3d_test(NAME TestE57 DATA bunnyInt19.e57 ARGS --load-plugins=pdal)
```

mwestphal commented on plugins/pdal/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+    EXTENSIONS e57
+    MIMETYPES model/e57
+    VTK_READER vtkPDALReader
+    FORMAT_DESCRIPTION "PDAL processes a variety of files: https://pdal.io/en/stable/stages/readers.html"
```
```suggestion
    FORMAT_DESCRIPTION "E57 Lidar Point Cloud Format"
```

mwestphal commented on plugins/pdal/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+  MIMETYPE_XML_FILES "${CMAKE_CURRENT_SOURCE_DIR}/f3d-pdal-formats.xml"
+  CONFIGURATION_DIRS "${CMAKE_CURRENT_SOURCE_DIR}/configs/config.d" "${CMAKE_CURRENT_SOURCE_DIR}/configs/thumbnail.d"
+  FREEDESKTOP
+)
```
missing end of line here

mwestphal commented on plugins/pdal/configs/config.d/10_pdal.json:
```diff
@@ -0,0 +1,9 @@
+      "load-plugins": "pdal"
+    }
+  }
+]
```
missing end of line here

RealEdwardS replied:
What's the significance of the end of line?

mwestphal commented on plugins/pdal/configs/thumbnail.d/10_pdal.json:
```diff
@@ -0,0 +1,9 @@
+      "load-plugins": "pdal"
+    }
+  }
+]
```
missing end of line here

mwestphal commented on plugins/pdal/pdal.inl:
remove this file

mwestphal commented on testing/data/DATA_LICENSES.md:
```diff
@@ -68,6 +68,7 @@
 - Wolf.fbx: OLEKSO: [CC-NC 4.0](http://creativecommons.org/licenses/by-nc/4.0/)
 - world\*: VTK Data: BSD-3-Clause
 - zombie\*.mdl: LibreQuake Project: BSD-3-Clause
+- bunnyInt19.e57: IOPDAL, e57 File: (http://www.libe57.org/data.html)
```
```suggestion
- bunnyInt19.e57: libE57.org: [MIT](http://www.libe57.org/data.html)
```

mwestphal reviewed (changes requested):
many changes needed :)

mwestphal commented:
Please resolve discussion you have adressed

mwestphal commented on .github/actions/proj-install-dep/action.yml:
```diff
@@ -0,0 +1,72 @@
+        cmake ../proj
+        -DBUILD_SHARED_LIBS=ON
+        -DCMAKE_BUILD_TYPE=Release
+        -DUSE_SQLITE:BOOL=ON
```
what about just disabling SQLITE ?

mwestphal commented on .github/actions/vtk-dependencies/action.yml:
```diff
@@ -60,3 +75,38 @@ runs:
+      uses: ./.actions/proj-install-dep
+      with:
+        cpu: ${{inputs.cpu}}
+        version: ${{inputs.proj_version}}
```
missing sqlite version here

RealEdwardS commented:
\ci fast

RealEdwardS commented:
\ci -fast

RealEdwardS commented:
\ci main

RealEdwardS commented:
\ci fast

RealEdwardS commented:
\ci -fast

RealEdwardS commented:
\ci main

mwestphal commented:
With 9.2.6 support removed, ill study how can this be integrated in a cross platform way.
